testplan/web_ui/testing/src/AssertionPane/AssertionTypes/TextAttachment.js

```diff
@@ -0,0 +1,270 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import {StyleSheet, css} from 'aphrodite';
+import axios from 'axios';
+import SyntaxHighlighter from 'react-syntax-highlighter';
+import { FadeLoader } from 'react-spinners';
+
+const DISPLAY_NUM = 20
+const SCROLLBAR_MIN_LIMIT = 60
+
+/**
+ * TextAttachment component:
+ *   * Gets text file using Axios
+ *   * Returns JSX depending on how long the text file is
+ */
+
+class TextAttachment extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+          loading: false,
+          error: undefined,
+          textContent: "",
+          originalText: undefined,
+          numberOfLines: undefined,
+          expandButtonPushed: false,
+          scrollbar: undefined,
+        };
+    }
+
+    /**
+    * Fetch the text attachment once the component has mounted.
+    * @public
+    */
+    componentDidMount() {
+        this.setState({loading: true});
+        this.getTextAttachment()
+    }
+
+    /*
+     * Prepare the contents of the collapsed .txt content, by displaying the
+     * last DISPLAY_NUM lines of the text file (or the entire text file if
+     * it has less lines than DISPLAY_NUM - short files).
+     *
+     * @param {array} text_array (each element of array is one line of text)
+     * @return {jsx} Content for the .txt file content when collapsed
+     * @private
+     */
+    getCollapsedText(text_array){
+        let last_lines;
+        let i;
+        let starting_line;
+        let num_of_lines = text_array.length;
+
+        if(num_of_lines < DISPLAY_NUM){
+            i = 0;
+            starting_line = 1;
+            last_lines = "";
+        }
+        else{
+            i = num_of_lines - DISPLAY_NUM;
+            starting_line = num_of_lines - DISPLAY_NUM;
+            last_lines = "...\n";
+        }
+
+        for(i; i < num_of_lines; i++){
+            last_lines = last_lines + text_array[i] + '\n';
+        }
+
+        let return_jsx = (
+            <SyntaxHighlighter showLineNumbers
+                startingLineNumber={starting_line}>
+              {last_lines}
+            </SyntaxHighlighter>
+        )
+
+        return return_jsx
+    }
+
+    /**
+    * Fetch the text attachment.
+    *   * Handle UID errors.
+    *   * Make a GET request for the .txt attachment file.
+    *   * Prepare the display content
+    * @public
+    */
+    getTextAttachment() {
+    const paths = window.location.pathname.split('/');
+
+    // Check if dev mode and display test_string
+    if ((paths.length >= 2) && (paths[1] === '_dev')) {
+        const TEST_TEXT = "test1\ntest2\ntest3\ntest4\ntest5\ntest6\ntest7\n";
+        let text = TEST_TEXT;
+        let lines = text.split('\n');
```

Comment:
It's a bit sad that we are duplicating logic between the dev mode and regular branches. I know it's not very straightforward because the regular mode is async. But maybe you can refactor into an extra helper function that takes the text string and handles setting the state from that - could call it something like `handleText(text)`. Then you can call that same helper function from the axios callback with the text data and also with the special dev test data.

Reply:
I saw there was quite a bit of code repetition, but I didn't think it was enough to refactor out, but turns out as there is logic repeated 3 times throughout the code, i'm now also convinced I should factor it out :D plus when it's right next to each other like that it looks even more horrible

https://stackoverflow.com/questions/2298272/how-much-duplicated-code-do-you-tolerate

Reply:
It is a question of judgement, sometimes a small amount of repetition can be OK if it helps readability - but this is the exception rather than the norm. In this case I think having a well-named function is more readable. A good rule of thumb is to think about what the duplicated lines are doing and if you can think of a good function name to describe what they do - if so that suggests you should factor them out into a function. A good function should mostly do one thing, and its name should tell you what that thing is.

In some cases the lines might do two (or more) distinct things, in which case you might consider factoring into two (or more) functions.

How far to take DRY is to a certain extent language dependent. In more verbose languages like C/C++, the "overhead" for creating a new function is non-trivial because you have to also create a function prototype in some other header file and worry more about collisions in the namespace, so developers tend to be lazier with allowing some repetition of a few lines here and there. In Python or JS the overhead of creating a new function is virtually nil, so there is less excuse for having even a couple of lines of repetition.